---
title: "Tutorial: Using the resources in Opal and DataSHIELD"
author:
- name: Juan R. Gonzalez
  affiliation:
  - &isglobal Bioinformatics Research Group in Epidemiolgy (BRGE), Barcelona Insitute for Global Health (ISGlobal)
  email: juanr.gonzalez@isglobal.org
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Tutorial: Using the resources in DataSHIELD}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(knitr)
BiocStyle::markdown()
options(width=100)
knitr::opts_chunk$set(comment="", warning=FALSE, message=FALSE, cache=TRUE)
```


# Getting started

This document can be reproduced by installing the following packages

```{r eval=FALSE}
install.packages("DSOpal")
devtools::install_github("datashield/dsBaseClient")
devtools::install_github("isglobal-brge/dsOmicsClient") 
```

Then the packages are loaded as usual

```{r load_packages}
library(DSOpal) 
library(dsBaseClient)
library(dsOmicsClient)
```




We have set up an [Opal demo site](http://opal-demo.obiba.org/) to illustrate how to perform some basic analyses using DataSHIELD as well as how to deal with different *resources* for 'omic data. The Opal server can be accessed with the credentials:

- username: `administrator`
- password: `password`


In this figure we can see all the projects available. 


```{r projects, echo=FALSE, fig.cap='Opal demo site available projects', fig.align='center'}
knitr::include_graphics("fig/opal_projects.png")
```

This tutorial will mainly make use of the resources available at `RSRC` project

```{r resources, echo=FALSE, fig.cap='Resources available at Opal demo site of RSRC project',  fig.align='center'}
knitr::include_graphics("fig/resources_rsrc.png")
```

In order to make the reader familiar with Opal we recommend visiting [the Opal online documentation](http://opaldoc.obiba.org/).



# Adding a new resource to the Opal server

The resources can be uploaded in the Opal server manually as described [here](https://isglobal-brge.github.io/resource_bookdown/tips-and-tricks.html#how-to-upload-a-new-resource-into-opal). However, it can also be done using R code as it is described in the following subsections. 

## Resource as a text file

Let us imagine that we have a `tsv` file that is stored in our hospital, server, cloud, GitHub repository or any other site. This file is containing information on several variables we want to analyze using DataSHIELD. Let us also imagine that this data is available at this URL: http://duffel.rail.bio/recount/TCGA/TCGA.tsv. This file encodes the phenotypes corresponding to 11,287 samples from [TCGA](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga) that are available in the [Recount project](https://jhubiostatistics.shinyapps.io/recount/). 

Thanks to the **resources** this dataset is not necessary to be uploaded into the Opal server as a table anymore. We can analyze this data with DataSHIELD packages by creating a new resource as following. 

Let us start by login the Opal server. **NOTE** that this requires full permissions and, hence, we access with administrator rights

```{r}
o <- opal.login(username = 'administrator',
                password = 'password', 
                url = 'https://opal-demo.obiba.org')
```

TCGA dataset can be added as a new resource as following

```{r}
opal.resource_create(opal = o, 
                     project = 'RSRC', 
                     name = 'pheno_TCGA', 
                     url = 'http://duffel.rail.bio/recount/TCGA/TCGA.tsv', 
                     format = 'tsv',
                     secret = 'GRGdder')
```



We can see that this resource have been added to our project by

```{r}
opal.resources(o, project='RSRC')
```


We can test the resource assignment. First we assign the resource to an object called `client` 

```{r}
opal.assign.resource(o, 'client', 'RSRC.pheno_TCGA')
opal.execute(o, 'class(client)')
```

We see that this object is of class `TidyFileResourceClient`. The [resourcer](https://cran.r-project.org/web/packages/resourcer/index.html) package will be use then to "resolve" this resource and to load it into the R server as we will see later. 



We logout the connection

```{r}
opal.logout(o)
```

Then, we can analyze the data using DataSHIELD by making use of the created resource. We start by login the resource using an user who have DataSHIELD permissions to our Opal server (dsuser). 

```{r}
builder <- newDSLoginBuilder()
builder$append(server = 'study1', url = 'https://opal-demo.obiba.org', 
               user = 'dsuser', password = 'password', 
               resource = 'RSRC.pheno_TCGA', driver = 'OpalDriver')
logindata <- builder$build()
```


Then, we load the resource into R as `res`

```{r}
conns <- datashield.login(logins = logindata, 
                          assign = TRUE, 
                          symbol = 'res')
```

The `resourcer` package which is installed in the Opal server contains functions that facilitates the access of this data in the R server. In particular, we will have access to the resource as a data.frame called `pheno`. To this end, `as.resource.data.frame ()` function is used to coerce the resource (e.g. `ResourceClient` object) to a data frame.

```{r}
datashield.assign.expr(conns, symbol = 'pheno', 
                       expr = quote(as.resource.data.frame(res)))
```

```{r}
ds.class('pheno')
ds.dim('pheno')
```

```{r}
datashield.logout(conns)
```


## Resource as a Rdata file (data frame)

Now, let us assume that our resource is a data frame saved in a .Rdata file available at: https://github.com/isglobal-brge/brgedata/raw/master/data/asthma.rda. We can do similar steps to load the resource into the Opal server:

```{r}

o <- opal.login('administrator','password', 
                url='https://opal-demo.obiba.org')

opal.resource_create(opal = o, 
                     project = 'RSRC', 
                     name = 'asthma', 
                     url = 'https://github.com/isglobal-brge/brgedata/raw/master/data/asthma.rda', 
                     format = 'data.frame')
     
opal.assign.resource(o, 'client', 'RSRC.asthma')
opal.execute(o, 'class(client)')

opal.logout(o)
```

In that case the class of the `client` object is `RdataFileResourceClient` whose resolver is also implemented in the `resourcer` package. 

Then, we can analyze this data using DataSHIELD functions as usual. In that case, instead of using `as.resource.data.frame ()` (it is also possible to use it) we make use of the function `as.resource.object ()` which coerce the resource to an internal data object that depends on the implementation of this object.

```{r}
builder <- newDSLoginBuilder()
builder$append(server = 'study1', url = 'https://opal-demo.obiba.org', 
               user = 'dsuser', password = 'password', 
               resource = 'RSRC.asthma', driver = 'OpalDriver')
logindata <- builder$build()

conns <- datashield.login(logins = logindata, 
                          assign = TRUE, 
                          symbol = 'res')

datashield.assign.expr(conns, symbol = 'asthma', 
                       expr = quote(as.resource.object(res)))
```


Now, we can perform some standard statistical analysis in DataSHIELD as usual:

```{r}
ds.class('asthma')

ds.colnames('asthma')

ds.glm(casecontrol ~ rs1422993 + smoke + bmi, data='asthma', family='binomial')
```

Do not forget to close the connection after finishing the analyses

```{r}
datashield.logout(conns)
```



## Resource as an Rdata file (ExpressionSet)

One of the main advantages of using the resources as an R file is that it may contains any type of R object. Let us illustrate this by having an R file with an `ExpressionSet` object which is a [Bioconductor] class to encapsulate omic data along with phenotypic information and annotation. See [here](https://kasperdanielhansen.github.io/genbioconductor/html/ExpressionSet.html) for a description and this figure for a visual idea of how these objects are organized 


```{r eSet, echo=FALSE, fig.cap='ExpressionSet infrastructure', fig.align='center'}
knitr::include_graphics("fig/eSet_vs_dataframe.png")
```

He have information on gene expression of 67,528 genes measured in 100 samples that are available in a `ExpressionSet` here: https://github.com/isglobal-brge/brgedata/raw/master/data/brge_gexp.rda. Let us illustrate how to create a resource in the Opal server having access to that data

```{r}
o <- opal.login('administrator','password', 
                url='https://opal-demo.obiba.org')

opal.resource_create(opal = o, 
                     project = 'RSRC', 
                     name = 'genexpr', 
                     url = 'https://github.com/isglobal-brge/brgedata/raw/master/data/brge_gexp.rda', 
                     format = 'ExpressionSet')

opal.assign.resource(o, 'client', 'RSRC.genexpr')
opal.execute(o, 'class(client)')
```

Now, we are ready to perform any data analysis using DataSHIELD as following. **NOTE** that in this case the R server must have installed those specific packages to deal with our specific class of objects. In that case, `ExpressionSet`s are managed with [`Biobase`](https://bioconductor.org/packages/release/bioc/html/Biobase.html) Bioconductor package. If so, the function `as.resource.object ()` which coerce the resource to an `ExpressionSet` that, in this case, will be available in the R server as an object called `eSet`.

```{r}
builder <- newDSLoginBuilder()
builder$append(server = 'study1', url = 'https://opal-demo.obiba.org', 
               user = 'dsuser', password = 'password', 
               resource = 'RSRC.genexpr', driver = 'OpalDriver')
logindata <- builder$build()

conns <- datashield.login(logins = logindata, 
                          assign = TRUE, 
                          symbol = 'res')

datashield.assign.expr(conns, symbol = 'eSet', 
                       expr = quote(as.resource.object(res)))
```

Then, we can use DataSHIELD fuctions as usual

```{r}
ds.class('eSet')
ds.dim('eSet')
```

In that case, if we execute

```{r error=TRUE}
ds.colnames('eSet')
```

we get an error, but ... we have developed a pair of DataSHIELD packages, called `dsOmics` and `dsOmicsClient` that allow us to deal with this type of objects in DataSHIELD. 

```{r}
dsOmicsClient::ds.featureData('eSet')
```

We finish the example by loggin out

```{r}
datashield.logout(conns)
```

# Using the resources in DataSHIELD 

## Data analysis combining different types of resources 

Here, we will use data from three studies that are available in our Opal demo repository. The three databases are called CNSIM1, CNSIM2, CNSIM3 and are available as three different resources: **mySQL database**, **SPSS file** and **CSV file** (see Figure \@ref(fig:resources)). This example mimics real situations where different hospitals or research centers manage their own databases containing harmonized data. Data correspond to three simulated datasets with different numbers of observations of 11 harmonized variables. They contain synthetic data based on a model derived from the participants of the 1958 Birth Cohort, as part of an obesity methodological development project. The available variables are:


```{r insert_table_variables, echo=FALSE}
vars <- readr::read_delim("fig/table_variables_cnsim.txt", delim=",")
kable(vars)
```

## Data analysis using a remote computation server


Let us finish our tutorial by removing the created resources

```{r}
opal.resource_delete(opal=o, project='RSRC', resource='pheno_TCGA')
opal.resource_delete(opal=o, project='RSRC', resource='asthma')
opal.resource_delete(opal=o, project='RSRC', resource='genexpr')
```


# Session Info

```{r}
sessionInfo()
```